### PR TITLE
Skip reconciliation on no-op claim and composite changes

### LINKED ISSUE #4703: Avoid no-op reconcilations
### What problem are you facing?

Resources are reconciliated on any change, but very it does not generate any change in the system, thus it could be easily skipped. The affected controller could be less busy, using less system resources.

### How could Crossplane help solve your problem?

Fine tune when a given reconcile loop is triggered:
* Claim reconciliation should be triggered only when (1) claim's annotation, labels an/or spec is changed, or (2) when composite status is modified
* Composite reconciliation should be triggered when (1) labels, annotations or spec is changed, (2) composition changes and/or (3) managed resource is updated

### PR DESCRIPTION
**Pull request by pedjak**
### Description of your changes

Claim reconciliation needs to happen only on:
* a non-status claim change and/or
* a non-metadata composite change for the given claim

Composite reconciliation needs to happen only on:
* a non-status composite change

Changes:
* Added `IgnoreStatusChanges` and `IgnoreMetadataChanges` predicates for proper event filtering

Fixes #4703 

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- ~[ ] Added or updated e2e tests.~
- ~[ ] Linked a PR or a [docs tracking issue] to [document this change].~
- ~[ ] Added `backport release-x.y` labels to auto-backport this PR.~

[contribution process]: https://github.com/crossplane/crossplane/tree/master/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/master/contributing#checklist-cheat-sheet

### CODE REVIEW

**Review by bobh66 — COMMENTED** [review]
This looks OK to me but I'd prefer to have an approval from someone who has better golang skills than I do.  Added @turkenh and @phisco to get some better-trained eyes on it.

File: pkg/controller/interceptor/interceptors.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+			d := time.Since(start).String()
+			if err == nil {
+				l.Info("Reconciliation done", "result", res, "duration", d)
+			} else {
+				l.Info("Reconciliation failed", "error", err, "duration", d)
+			}
+		}()
+		return r.Reconcile(ctx, req)
+	}
+}
```
Comment (phisco) [inline pkg/controller/interceptor/interceptors.go:47]: We currently don't log much at `Info` level, so I think it would make more sense to at least lower the level here to `Debug` or if you feel this needs to be at `Info` level, please start a conversation to change the project's approach to logging (if there is no dedicated issue already) and move this to a dedicated PR, so that it doesn't hold up the rest of the PR 🙏 WDYT @pedjak?

Comment (sttts) [reply pkg/controller/interceptor/interceptors.go:47]: and please split into two commits. This is not one change.

Comment (pedjak) [reply pkg/controller/interceptor/interceptors.go:47]: > We currently don't log much at Info level, 

True,  we do it in a handful of places: https://github.com/search?q=repo%3Acrossplane%2Fcrossplane%20%22.Info(%22&type=code

Given that this PR is about lowering the number of reconcile rounds, I thought it would be good also to make that trackable even if debug logs are not turned on, hence this change here.

Having single info line for each reconciliation improves UX, i.e. user get a notion what is going on, i.e. what gets reconciled. Today this mostly happens silently and it is super hard to diagnose whether it happens and why.

Comment (phisco) [reply pkg/controller/interceptor/interceptors.go:47]: We do expose prometheus metrics with similar, although less granular, [metrics](https://book.kubebuilder.io/reference/metrics-reference). But, I do agree we could benefit from being more verbose, I just feel that's a major policy change that should be discussed more broadly.

Comment (pedjak) [reply pkg/controller/interceptor/interceptors.go]: `LoggingInterceptor` moved into the separate PR: #5143.

**Comment by negz** [comment]
I see that @sttts had some concerns about doing this per https://github.com/crossplane/crossplane/issues/4703#issuecomment-1746809710. @sttts can you elaborate? My guess would be you're worried about false negatives - i.e. we skip some reconciles that were actually useful in some way we didn't expect?

If we did want to build confidence, we could use a feature flag.

**Comment by sttts** [comment]
Yes, skipping reconciles that matter. It complicates the state machine.

Metrics are nice, but how often does anybody look at metrics? Especially us devs. We might eventually notice that Crossplane is heavy but have no clue why. Metrics don‘t tell a helpful story.
